pep-9999.rst

```diff
@@ -0,0 +1,158 @@
+  confusion for packagers: license-related `Classifiers` and the `License`
+  attribute.
+
+- Furthermore, it is difficult to convey anything but simple licensing using
+  these two fields. License information expressiveness and clarity when creating
+  packages will help packagers and users alike.
+
+License expressions provide a better way to express the license of a
+distribution using a well defined syntax and well known license ids from SPDX.
+
+This PEP was started based on several discussions and in particular:
+
+- `<https://github.com/pypa/warehouse/issues/2996>`_
+- `<https://github.com/pypa/interoperability-peps/issues/46>`_
+- `<https://github.com/pypa/packaging-problems/issues/41>`_
+- `<https://github.com/pypa/wheel/issues/138>`_
+- `<https://github.com/pombredanne/spdx-pypi-pep/issues/1#issuecomment-281343077>`_
+
+
+Fields
+======
+
+The canonical source for the names and semantics of each of the supported
+metadata fields is the `Core Metadata Specification`_ document.
+
+The details for the update to the `Core Metadata Specification`_ document are 
+in this `Core Metadata Specification pull request`_.
+
+
+New in Version 2.2
+------------------
+
+License-Expression (optional)
+:::::::::::::::::::::::::::::
+
+A string containing a license expression using the SPDX license expression
+syntax and SPDX-listed license identifiers so that user can express accurately
+the licensing of a package in a way that is both easy to read for humans and
+easy to process automatically by tools.
+
+A string containing a license expression using `SPDX license expression syntax and
+identifiers <https://spdx.org/ids>`__ that indicates the license(s) covering the
+distribution. The license expressions are specified in the `SPDX specification 
+<https://spdx.org/spdx-specification-21-web-version#h.jxpfx0ykyb60>`_.
+SPDX also publishes a companion `list of license ids <https://spdx.org/license-list>`_
+for use in license expressions. 
+
+Note: this field is not case sensitive though the normalized representation of
+an SPDX license expression has a well defined case.
+
+Examples::
+
+    License-Expression: BSD-3-Clause
+
+    License-Expression: MIT OR GPL-2.0-or-later AND (FSFUL AND BSD-2-Clause)
+
+    License-Expression: GPL-3.0-or-later WITH Classpath-Exception-2.0 OR BSD-3-Clause
+
+
+Tools `should` validate that this field contains a valid license expression
+and `should` report a warning if the field contains an invalid expression.
+
+
+Changed in Version 2.2
+----------------------
+
+License
+:::::::
+
+This field will be deprecated in the future and its documentation has been
+updated accordingly. Tools are encouraged to provide a warning when this field
+is used.
+
+Classifiers
+:::::::::::
+
+The use of license-related classifiers in this field will be deprecated in the
+future and its documentation has been updated accordingly. Tools are encouraged
+to provide a warning when this field is used with license-related classifiers.
```

Comment:
I don't like the idea of deprecating these fields, as SPDX is driven by the needs of large consumer organisations, and if they're not offering to pay project maintainers to update their licensing metadata (e.g. through a Tidelift subscription or a consulting contract), then it isn't reasonable to push new work onto publishers solely for the benefit of these organisations.

Instead, I'd prefer to see these sections say something along these lines:

License:

> If the License-Expression field is present, publishing tools MUST NOT also populate the License field. However, for compatibility with existing publishing and installation processes, the License field SHOULD continue to be accepted if the License-Expression field is absent. Publishing tools MAY infer License-Expression from the provided License information if they are able to do so unambiguously.

Classifiers:

> If the License-Expression field is present, publishing tools MUST NOT also provide any licensing related Classifiers entries. However, for compatibility with existing publishing and installation processes, licensing related Classifiers entries SHOULD continue to be accepted if the License-Expression field is absent. Publishing tools MAY infer License-Expression from the provided Classifiers entries if they are able to do so unambiguously.

> However, no new licensing related classifiers will be added, with anyone requesting them being directed to use the License-Expression field instead.

Reply:
Note that omitting the deprecations doesn't actually make things any more complicated for metadata consumers, since they have to deal with at least the 1,428,826 project releases already on PyPI anyway, and none of those will have the License-Expression field.

The important part of this PEP is providing a way for folks that already care to be unambiguous about their licensing, and to offer a low impact migration path if they want to send PRs to other open source projects that they would also like to see clarify their licensing.

Reply:
I'll second @ncoghlan.

Let's not make existing packages incompatible with Metadata 2.2 and instead, add `License-Expression` as an opt-in unambiguous alternative to existing mechanisms. Having it be exclusive to License Classifiers and "License" metadata is a nice approach for that.

Reply:
@Conan-Kudo :wave:  Thank you for reviewing this!

you wrote:

> I don't generally agree that there are significant benefits to regular people to mandate SPDX. 

We are not mandating SPDX here. Instead suggesting clarity in the license declarations with warnings and suggestions. A list of known licenses is very much what we have today with License classifiers by the way, but that is limited and ambiguous unless you have a single license. This list of Classifier is also outdated in general (and this was one of the triggers for this PEP based on @Steap tickets. 

> Additionally, I don't think it's worth it to repurpose the existing License field.
> If people want to fill in this field, then adding a SPDX-License-Expression tag makes sense.

Well if we add a new optional field, we are making the spec much heavier but that also means that it will mean more things to deal with for packagers and users. If find it much more efficient to say that we are going to start soft validating what you put in the License field so it is a tad a more formally defined. A new field will IMHO bring some confusion and will most likely not be much used. And if this is not used, then why bother at all adding it?

> Though, I'm tempted to suggest that you make this a versioned tag (like SPDX-3.0-License-Expression) because SPDX does not remain consistent on license tags.

My gut reaction is that this would be even more complex and more confusing than using just an extra field. In practice the SPDX is a slow moving thing and I do not think we need the complexity of tracking exactly the license list versions. Except for one past event in late 2017 when the FSF licenses ids where updated (at the request of rms and the FSF) things are evolving slowly as they should. Therefore, I do not think that there is any use case that is so critical that it would be worth tracking the list versions.

Reply:
@pombredanne The problem with SPDX is its inconsistent philosophy. You're more-or-less okay with having a single `MIT` reference to represent dozens of MIT variants, while simultaneously generating useless BSD variants (and not even enough of them). Go one way or the other. Personally, I would have preferred the Fedora-style (as SPDX did with MIT licenses) as opposed to Debian-style (as SPDX is attempting to do with BSD licenses).

And the problem with SPDX is that it only works as long as the license has an identifier. And SPDX is far from responsive for solving this problem. 

One of the core reasons why Fedora hasn't adopted SPDX license identifiers is because it doesn't solve the problem, and the SPDX group doesn't appear to want to.

Let's suppose SPDX decides to do more categorization of BSD/MIT variants and realizes there needs to be a format change to go with it (there's at least the vendor suffix thing under consideration for dealing with copyright variants, which is going to a whole new level of stupid), then all this stuff becomes invalid again.

I've already dealt with three iterations of incompatibilities as SPDX license identifiers and syntax changed in openSUSE, and it was not fun having stuff just suddenly be invalid like that.

In my view, I don't think it's worth it to make it part of Python unless there's some resolution to these problems. And many of these problems are also the hiccups to adopting them in Fedora.

Reply:
@Conan-Kudo you wrote

> The problem with SPDX is its inconsistent philosophy. You're more-or-less okay with having a single `MIT` reference to represent dozens of MIT variants, while simultaneously generating useless BSD variants (and not even enough of them). Go one way or the other. Personally, I would have preferred the Fedora-style (as SPDX did with MIT licenses) as opposed to Debian-style (as SPDX is attempting to do with BSD licenses).

I am not sure there is a specific philosophy adopted there. The team of volunteer lawyers that deals with this @ SPDX is trying to handle it such things are legally correct and avoid license ids inflation when possible. And they have also tried to reconcile historical approaches of Fedora and Debian. That said, this is all open... you should chime in there to voice your concerns IMHO.

> And the problem with SPDX is that it only works as long as the license has an identifier. 

There are LicenseRef when using documents and there is an emerging way that is specified to have "namespaces" for ids. 

> And SPDX is far from responsive for solving this problem.

SPDX are people: this is a bunch of volunteers with a finite amount of time like we all are!
So i you care about this and think this is a problem, you should join the fray and lend a hand IMHO.

> One of the core reasons why Fedora hasn't adopted SPDX license identifiers is because it doesn't solve the problem, and the SPDX group doesn't appear to want to.

I am not sure which is the "the problem" you are talking about... but that said, this is something that you should really bring up to the SPDX project. It would be great to Fedora also use SPDX.

> Let's suppose SPDX decides to do more categorization of BSD/MIT variants and realizes there needs to be a format change to go with it 

Can you elaborate a it? what would be a format change?

>(there's at least the vendor suffix thing under consideration for dealing with copyright variants, which is going to a whole new level of stupid), then all this stuff becomes invalid again.

I am not sure I know what this is about. Is this about license namepaces?
 
> I've already dealt with three iterations of incompatibilities as SPDX license identifiers and syntax changed in openSUSE, and it was not fun having stuff just suddenly be invalid like that.

Is this about the "-only" vs. "-or-later" suffixes changes in the FSF licenses id in late 2017? If this is the case, then I am with you there: that change could have been avoided, but the FSF seemed to care a lot about it. Then again you should bring your concerns to the SPDX mailing lists alright. This is the only way to be heard! 

> In my view, I don't think it's worth it to make it part of Python unless there's some resolution to these problems. And many of these problems are also the hiccups to adopting them in Fedora.

I appreciate your concerns but I am not sure I get exactly what the problems are. Can you elaborate them a tad more to make sure I can understand your arguments and provide a proper response to these objections?

Reply:
@Conan-Kudo you also seem to flat out reject any usage of SPDX licenses ids and expressions, yet what would you suggest as an alternative to the problematic status-quo?

Reply:
@ncoghlan It is true that we do not want to force package maintainers to do additional work, but the point is to make it easier for downstream package maintainers (in Debian, Fedora, *BSD...) to know what license is used (and automate the process). I assure you I did not have "large consumer organisations" in mind when I asked for more precise classifiers.

@Conan-Kudo: SPDX seems like a de-facto standard to me. It is also used by other projects. I second @pombredanne here: what alternative would you suggest? I'm not against using something other than SPDX, but we would need to have something less ambiguous than the current classifiers, and that would be able to evolve over time.

Reply:
@pombredanne, @Steap: The two systems of license identifiers I'm generally okay with are [Fedora's](https://fedoraproject.org/wiki/Licensing:Main#Software_License_List) and SPDX. However, for SPDX, I don't think it's sane or safe to suggest we should just reference SPDX's list of identifiers. No matter what system is chosen, a _locally versioned copy_ should exist so that validation can be precisely done based on what we expect.

And if we want license expressions to work, it should be a brand new field because then tools can reasonably expect the data in that field to be formulaic and sane to process.

The existing set of Trove classifiers are obviously not complete enough, but I loathe to make it so we're reliant on an unstable validation target. My experience dealing with this in openSUSE was sufficiently painful that my trust in SPDX to not do this again is mostly gone.

Reply:
@pombredanne Between the times that the expression grammar was changed (`+` -> `WITH`, `/` -> `OR`, etc.), the stupid case sensitivity issue (`and` vs `AND`), and the change of the GNU license tags (`GPL-2.0` -> `GPL-2.0-only`, `GPL-3.0+` -> `GPL-3.0-or-later`, etc.), I've lost faith in the SPDX organization to keep this stable and reliable.

openSUSE, Rust, and the Linux kernel all implement SPDX license identifiers differently based on when they pulled the rules. And that's the most frustrating part of it all! Suddenly many things failed validation when they used to pass because the tools and data were updated to invalidate them.

I don't want that for Python, at all.

Reply:
@Conan-Kudo You wrote in https://github.com/pombredanne/spdx-pypi-pep/pull/2#discussion_r321946465 

> The two systems of license identifiers I'm generally okay with are Fedora's and SPDX. However, for SPDX, I don't think it's sane or safe to suggest we should just reference SPDX's list of identifiers. No matter what system is chosen, a locally versioned copy should exist so that validation can be precisely done based on what we expect.

So you are saying we should store the list of ids @ Python, (much like we do with the list of classifiers today) and not depend on an external source of ids? 
That could work for sure: I have taken a slightly different approach so far here by only specifying an SPDX license list version (and any compatible future version), which is IMHO both precise and flexible enough to make this PEP future prrof and leave the list of actual Ids being used to the packaging documentation (the same way updating the Classifiers does not require a new PEP today: a simple patch is enough)
We could be explicit in the packaging documentation about which exact version of the SPDX license list we support at any time. Would this work out for you?


> And if we want license expressions to work, it should be a brand new field because then tools can reasonably expect the data in that field to be formulaic and sane to process.

This is something that has been discussed a bit here and also at https://discuss.python.org/t/improving-license-clarity-with-better-package-metadata quite a bit. There have been opinions going one way (reuse the License field) or the other (add a new field). As you say here, there are direct benefits to a new field. The drawbacks IMHO are that this is a new field and that this requires cross-field validation and make some fields mutually exclusive in some cases based on their content. Since we do not want to make this mandatory, I see a new field as being much more invasive than reusing the existing `License` field and gently nagging users to recommend using an license expression there.

I am not strictly opposed to another field but at the moment there has been no strong objections to reusing License except for yours, so this is tracked as one of the rejected ideas. 


> The existing set of Trove classifiers are obviously not complete enough, but I loathe to make it so we're reliant on an unstable validation target. My experience dealing with this in openSUSE was sufficiently painful that my trust in SPDX to not do this again is mostly gone.

I really would like to understand the specific of this bad experience you have had with openSUSE! Can you explain more?

Reply:
@Conan-Kudo you wrote in https://github.com/pombredanne/spdx-pypi-pep/pull/2#discussion_r321946825:

> @pombredanne Between the times that the expression grammar was changed (+ -> WITH, / -> OR, etc.),

From what I can see there never was any such change: the expression grammar was introduced in SPDX 2.0 and there are no material changes in version 2.1. I extracted the texts from each version for the Expression spec part and posted the two versions there https://gist.github.com/pombredanne/49b2b8699d15403bec21030fe359c797/revisions

- `+` is not replaced by `WITH`: both were always there
- `/` was never a substitute for `OR`: you may refer too the way this has been used in Rust Cargo manifests... but that's Rust way and this was never specified in SPDX specs.


> the stupid case sensitivity issue (and vs AND)

Being a Pythonista I am with you there: everything should be in proper lower case and snake case! :D

That said that's minor since this is only a canonical form issue. In practice you can ignore case when parsing (e.g. being lenient on inputs) and output something with the proper case. The license-expression library does exactly this FWIW, so I do not see this as a problem.


> and the change of the GNU license tags (GPL-2.0 -> GPL-2.0-only, GPL-3.0+ -> GPL-3.0-or-later, etc.), 


I am siding with you there too. And I voiced my concerns publicly when that changed happened in the fall of 2017.  I was against it, but I was not the majority so I accepted the community consensus. You should have voiced your concerns in pblic then ... I would have mucho appreciated the support. In the end the SPDX community showed deference to rms and the request of the FSF to use these updated ids for their own licenses. Even though I was against it (especially since I was in the middle of a major licensing clarification work with the linux kernel maintainers), when a license author (and someone as prominent in the community as rms himself) comes and kindly ask for changes on how their things are named, I think this is OK to do the change. If you still disaagree, please bring your concern to the SPDX mailing list, as well as to rms and the FSF, but that ship has sailed now IMHO.


> I've lost faith in the SPDX organization to keep this stable and reliable.

I do not recall that you joined the discussions at SPDX back then (but I have a crass memory, so forgive me if this is wrong), and that's really something that I would have appreciated as we are thinking alike on many of these topics and you seem to care about these.

That said, things are versioned and evolve in SPDX, they are not set in stone! The same way I expect software to evolve with bug fixes and new features. To the credit of the volunteers caring for the SPDX spec and license list, things are versioned and they are trying hard to keep backward compability AND ensure that new and retired license ids are never reused and have a clear mapping. I think that's quite OK in general. I am not sure what else I could


> openSUSE, Rust, and the Linux kernel all implement SPDX license identifiers differently based on when they pulled the rules. And that's the most frustrating part of it all! Suddenly many things failed validation when they used to pass because the tools and data were updated to invalidate them.

I can understand your frustration, but SPDX spec and ids list are versioned for a good reason to cope with changes. You cannot blame tools that may not work yet with newer versions of the specs or stop to support previous versions. This is no different from code: there are at times some major changes and they may break compatibilty. That said, tools can handle that alright: for instance the license-expression library that I maintain is prefectly happy with past, current and future versions of the SPDX license list as well as mixing all versions together(and FWIW with any list of license symbols you can feed it with) so this means that it is possible for a tool to deal with updates in an orderly way.

That said, if I dive in the specifics:

- For openSUSE, I can see the guidelines here https://en.opensuse.org/openSUSE:Packaging_guidelines#Licensing and they may be a tad underspecified as this references an SPDX license "short names" as opposed to an "identifier" but that's quite minor.

There is a also an mapping table at https://docs.google.com/spreadsheets/d/14AdaJ6cmU0kvQ4ulq9pWpjdZL5tkR03exRSYJmPGdfs/pub which looks to me as bringing order to map several legacy openSUSE ids to a sinegle SPDX id such as for AGPL... and I see that as a good thing. Now you seem much closer than I am to openSUSE so you likely know better.

```
AGPL-3.0-only	Affero GPL
AGPL-3.0-only	AGPL-3.0
AGPL-3.0-only	AGPLv3
AGPL-3.0-or-later	AGPLv3+
AGPL-3.0-or-later	AGPL-3.0+
AGPL-3.0-or-later	SUSE-AGPL-3.0+
```

- Rust is very clear on what they use and that sounds very clean and consistent to me.
@wking @dwijnand If could comment on the Rust/Cargo adventure with SPDX license expressions that would be awesome!

See https://doc.rust-lang.org/cargo/reference/manifest.html#package-metadata
And https://github.com/rust-lang/cargo/blob/fe0e5a48b75da2b405c8ce1ba2674e174ae11d5d/src/doc/src/reference/manifest.md#L254
And https://github.com/rust-lang/cargo/blame/fe0e5a48b75da2b405c8ce1ba2674e174ae11d5d/src/doc/src/reference/manifest.md#L254

```
# This is an SPDX 2.1 license expression for this package. Currently
# crates.io will validate the license provided against a whitelist of
# known license and exception identifiers from the SPDX license list
# 2.4. Parentheses are not currently supported.
#
# Multiple licenses can be separated with a `/`, although that usage
# is deprecated. Instead, use a license expression with AND and OR
# operators to get more explicit semantics.
license = "..."
```

- For the Linux kernel, I have been in the first line as I helped Greg Kroah-Hartman, Thomas Gleixner and Kate Stewart to streamline the kernel licensing by using SPDX expressions in source code and the FSFE reuse conventions. The bulk of the initial work took place in fall 2017 just when the FSF requested the GPL id cange @ SPDX... this has been handled nicely and without much hiccups by pinning the version of the licenses being used. There are tons of discussions on lkml on this and many patches since thousands of files were changed, but tell me if you want some pointers on specifics. 

That said you may be referring to things like this thread https://lists.opensuse.org/opensuse-factory/2018-02/threads3.html https://lists.opensuse.org/opensuse-factory/2018-02/msg00464.html

As I mentioned here, I was first line when that happened and I am with you on the issue. I explained here why this happened with rms and the FSF. And I was against it, but that was not te consensus. It would have been great for folks from openSUSE to chime in at the ttime.



> I don't want that for Python, at all.

I highly respect your opinions there: since you care and want something different for Python, may I kindly suggest that you submit a PR on top of this PR/branch with your suggested updates or specific comments to evolve it?

Or if you think this PEP is completely off base and wrong, would you mind to start a concurrent PEP with an alternaive proposal so that we can all discuss and review both proposals to work out something constructive?

Reply:
@pombredanne I think it's perfectly fine to allow people to put SPDX expressions in `License`, but adding a field that can be optionally used that _must_ be SPDX compliant means that if that field is detected, software can guarantee they can process it that way.

I'm aware that the identifiers list is actually versioned. What I'm saying is that this PEP should specify a version and have a local copy of it, and updating the version of the identifiers should require a revision to this PEP.

And for what it's worth, I have been subscribed to the main SPDX mailing list (I subscribed when I was more interested in migrating Fedora to SPDX identifiers as part of my Fedora Rust SIG work), and I voiced my concern about the change when it happened as well. That was the straw that broke the camel's back for me, as the discussion did not resolve the issue well for me. I would have probably been less annoyed if it was easy to directly request specific versions of the identifier list for machine parsing.

Reply:
@Conan-Kudo thanks for your answer details... I was afk for a while and I am back.

You wrote in https://github.com/pombredanne/spdx-pypi-pep/pull/2#discussion_r321973682:

> I want updating to new SPDX identifier versions to require PEP updates.

That's reasonable (this is more or less what we ended up doing for the Linux kernel: we store the list of valid identifiers in the kernel doc)... but at the same time, this means an update to a foundational document (the metadata doc/PEP) which is fairly significant and not to be done lightly.

To me the key thing would be how often would this possibly happen in the future? The rate at which new FOSS and related licenses evolve is slow enough. Here are some anecdotes:

1. SPDX adds about 20 new licenses per year
2. In ScanCode this is more like 1 or 2 per week.

That said, the new added licenses are mostly either old, seldom used licenses that were not yet "discovered" and new licenses that are not much used as of now. So in 99% of the cases the new licenses could be qualified as exotic.

Therefore, I think we could freeze a version of the SPDX license list in the PEP alright and the need for an update should be rather rare (maybe once in three years or so). 

@Steap @ncoghlan @pfmoore  @cjerdonek  what would you think about this? This would mean being strict in this section: https://github.com/pombredanne/spdx-pypi-pep/pull/2/files#diff-7a25ca1769914c1141cb5c63dc781f32R223 and specify that we use a defined version of the list and that adopting future version would require an update to the metadata doc and version.

- The positive: there is no ambiguity about which licenses ids are supported
- The negative: adopting a new version of the license list every other year would require a new PEP, which is a disruption but that would be every couple years or so only.

Reply:
@Conan-Kudo you wrote in https://github.com/pombredanne/spdx-pypi-pep/pull/2#discussion_r321973682

> I think it's perfectly fine to allow people to put SPDX expressions in `License`, but adding a field that can be optionally used that _must_ be SPDX compliant means that if that field is detected, software can guarantee they can process it that way.

My personal inclination as documented in the current version is to avoid field inflation and reuse the license field. There is no ambiguity at all when this contains a parsable SPDX license expression or not. Since the field is in use and a warning would be issued it provides the proper gentle nagging that will help authors evolve towards a more accurate license documentation. a field that's new and optional is likely to have a lower impact and create a bigger disruption:

We have today two fields used for license (and this is confusing). And we would go to three fields all optional if we add a new one, a likely source of more confusion for authors IMHO.

With that said, if there is a consensus to use a separate field, I will update the draft to use that instead. 

@Steap @ncoghlan @pfmoore @cjerdonek @pradyunsg what's your last take on this topic?

This and the freezing the list of licenses discussed in https://github.com/pombredanne/spdx-pypi-pep/pull/2#discussion_r330419938 are IMHO the last two objections/concerns to address and resolve before moving this PEP to an official draft IMHO

Reply:
Let's do these discussion in dedicated issues for each of these points? It'd be weird to discuss these inline on a PR.